server/mqtt.go

```diff
@@ -2336,10 +2336,17 @@ func (as *mqttAccountSessionManager) serializeRetainedMsgsForSub(sess *mqttSessi
 			sub.mqtt.prm = &mqttWriter{}
 		}
 		prm := sub.mqtt.prm
-		pi := sess.getPubAckIdentifier(mqttGetQoS(rm.Flags), sub)
+		var pi uint16
+		qos := mqttGetQoS(rm.Flags)
+		if qos > sub.mqtt.qos {
+			qos = sub.mqtt.qos
+		}
+		if qos > 0 {
+			pi = sess.trackPublishRetained()
```

Comment:
So in this case we know that we should get a `pi > 0` (in that it is QoS > 0 for both message and subscription). Here I wonder if `pi == 0` if we should not skip delivery of that retained message - since we can't find a suitable PI. Not sure of the implications there, and maybe going with PI==0 is better than nothing, but just wanted to raise the issue.

Reply:
Hmmm, good point @kozlovic. Passing PI 0 down the line is not a good idea anyway, may be a place in the QOS code where it is assumed to be non-0. Maybe we could (should?) set the (delivery) `qos` to 0 for that message? That'd be a 1-line change...

Reply:
See https://github.com/nats-io/nats-server/pull/4349/commits/a2452aa844ff4609728a223c3c76d7d67927d922